### ISSUE TITLE
Why are MRET and SRET being extended to add memory fences?

### ISSUE THREAD

**Issue by kdockser** [riscv/riscv-svvptc#9]
In order to eliminate the need to execute an SFENCE.VMA when enabling a page table entry, we are modifying the semantics of MRET and SRET to include fencing functionality for implicit memory accesses similar to that of the SFENCE.VMA. Not only have we not eliminated such fencing, but we are now requiring it for all MRETs and SRETs, even when not needed.

The original idea of Svvptc was to "simply" disallow holding invalid PTEs in the address-translation cache.  Now there seems to be no such requirement. In its absence, it seems that the new fencing functionality of MRET/SRET would also need to _invalidate entries in the address-translation cache_ to cover this case. If so, how does the MRET/SRET know what to invalidate?

Taking a step back, it seems like we are trying to accomplish two things:
1. Prevent invalid PTEs from being held in the address-translation cache (ATC). Or take a trap on encountering such an entry in a PTW so that it can be flushed with an appropriate SFENCE that targets the invalid entry.
3. Allow a speculative page table walk for a storage instruction that reads an invalid PTE to retry once this storage instruction is the oldest uncommitted. This could be done by HW or by taking a trap (which executes an SFENCE).
It seems like we can come up with a viable architectural solution with these two components that doesn't involve placing an undue burden on the *RET instructions.

**Comment by ved-rivos** [comment]
The specification is not described in form of address translation caches is because ATC are not an architectural construct. The specification is not adding additional semantics for SRET or MRET architecturally. The specification is stating the boundary where software can expect a store to make a PTE valid can be observed. This is to avoid placing requirement on implementations to need to observe a store to a PTE by the very next instruction that performs the store. The boundary is defined as a naturally synchronizing point such as an SRET, MRET, or a memory management fence.

**Comment by kdockser** [comment]
Thanks for your detailed reply, Ved.

I believe that this specification _is_ changing the semantics of *RET instructions in that it is requiring a *RET instruction to act as an effective memory-management fence. Either the *RET would need to be involved in the fencing behavior itself, or there would need to be another means to independently provide this fencing behavior quickly enough that it would give the same effect as if it were produced by the *RET.

In any case, I think the wording of the specification can be a little hard to parse. How about something a little more concise such as:

`When the Svvptc extension is implemented, the execution of an SRET or an MRET instruction ensures that any previous store in a hart that updates the Valid bit of a PTE from 0 to 1 is ordered before implicit references by subsequent instructions in that hart to the memory-management data structures.`

**Comment by kdockser** [comment]
Since the purpose of this extension is to allow SW to elide the use of an SFENCE after making a PTE valid, it is necessary that this extension prohibits non-valid PTEs from being cached in the Address Translation Cache (ATC). If the non-valid PTE is cached in the ATC, it _will_ be necessary to invalidate the TLB entry.

While ATCs are neither mandated nor fully specified by the architecture, the Privileged Architecture _does_ define the architectural behavior of ATCs including its behavior in the Virtual Address Translation Process. Therefore it is perfectly valid --- indeed, it is essential --- for the Svvptc extension to explicitly state any requirements that it imposes on the ATC.

That said, no such change needs to be made to the normative text if the intention is for implementations of Svvptc to be able elide SFENCEs even if the ATC caches non-valid PTEs. If that is the case, this should at least be made clear in the non-normative text as it might not be obvious to implementers.

**Comment by ved-rivos** [comment]
@kdockser thanks. I am in process of incorporating your suggestions and a few other updates in the document. I will provide an update here shortly.

**Comment by ved-rivos** [comment]
@kdockser - Please take a look at the updated [text](https://github.com/riscv/riscv-svvptc/releases/download/v0.9/riscv-svvptc.pdf). I hope this addresses these questions.

**Comment by kdockser** [comment]
Thanks @ved-rivos, that is lot clearer and a lot less onerous.

**Comment by ved-rivos** [comment]
Thanks.
